Allow instantiation of TriplesFactory without label-to-id mapping

mberr opened the pull request:
This PR implements parts of #57 

In particular it adds a new constructor to `TriplesFactory.create_without_labels`, and makes passing label-to-id mappings to the base constructor of `TriplesFactory` optional. When accessing methods / attributes which require labels, an exception is raised.

mberr committed:
allow instantiation without label mapping

mberr committed:
Fix init and type annotation

mberr committed:
Add with_labels method

PyKEEN-bot committed:
Trigger CI

mberr committed:
Merge remote-tracking branch 'origin/master' into premapped-triples

mberr committed:
Fix docstring

PyKEEN-bot committed:
Trigger CI

mberr committed:
mypy fixes

mberr committed:
Merge remote-tracking branch 'origin/premapped-triples' into premapped-triples

mberr committed:
extract mapping

mberr committed:
extract base class

mberr committed:
Add docstrings

mberr committed:
Remove unused properties

mberr committed:
Fix missing docstrings

mberr committed:
start moving functionality to CoreTriplesFactory

mberr committed:
update docstring

mberr committed:
Move ID-based methods to CoreTriplesFactory

mberr committed:
fix infinite recursion

mberr committed:
fix some issues

mberr committed:
Merge branch 'master' into premapped-triples

PyKEEN-bot committed:
Trigger CI

mberr committed:
Fix mypy

mberr committed:
Fix test

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Merge branch 'master' into premapped-triples

cthoyt committed:
Update type checking

cthoyt reviewed (changes requested):
I like the use of subclassing for CoreTriplesFactory and TriplesFactory that maintains the old interface. However, in places where the subclass isn't necessary, we should update the type hints to reflect that only a CoreTriplesFactory is necessary.

Besides that, I think we should provide a couple more test scenarios for core triples factory

cthoyt commented on src/pykeen/triples/triples_factory.py:
```diff
@@ -146,59 +146,465 @@ def _ensure_ids(
+        :return:
+            A new triples factory.
+        """
+        if num_entities is None:
```
why are these two none checks not directly in `__init__`?

mberr replied:
It enforces the user either to explicitly provide `num_entities` / `num_relations`, or use the `create` factory method.

cthoyt commented on src/pykeen/triples/triples_factory.py:
```diff
@@ -146,59 +146,465 @@ def _ensure_ids(
+        relation_to_id: Mapping[str, int],
+    ) -> "TriplesFactory":
+        """Add labeling to the TriplesFactory."""
+        # TODO: Check labelings
```
can you elaborate on what needs to be done here? We can push it to another PR but at least write some more so we don't forget

mberr replied:
The mappings should be compatible to the triples, i.e. their value range cover at least all occurring entity/relation IDs.

mberr replied:
It should be an easy fix:

```python
set(self.mapped_triples[:, [0, 2]].unique().tolist()).issubset(entity_to_id.values())
```

cthoyt replied:
Then let's put it in this PR

mberr replied:
Done in 6054267

cthoyt committed:
Flip logic to make less nested

mberr committed:
Implement labeling verification for CoreTriplesFactory.with_labels

cthoyt committed:
Update exports

cthoyt committed:
Reimplement triples factory generation with core triples factory

cthoyt committed:
Abstract functionality that can be performed by core triples factory

@mberr need a design decision on how a core triples factory should store the entity and relation ids

cthoyt committed:
Add tests demonstrating using the pipeline with core triples factory

however the whole prediction pipeline will not work right now becuase it requires labeled triples factory functionality

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Update test_model_mode.py

Trigger CI

cthoyt commented on src/pykeen/triples/triples_numeric_literals_factory.py:
```diff
@@ -85,6 +85,7 @@ def __init__(
         elif path_to_numeric_triples is not None:
             numeric_triples = load_triples(path_to_numeric_triples)
 
+        assert self.entity_to_id is not None
```
is this assert needed?

cthoyt committed:
Merge branch 'master' into premapped-triples

cthoyt committed:
Update __init__.py

Trigger CI

cthoyt reviewed (approved):
Next PR we can move more functionality to the base class

mberr committed:
Merge branch 'master' into premapped-triples

PyKEEN-bot committed:
Trigger CI
